File: src/e1palindromepermutation/CS/PalindromePermutation.Algo/PalindromePermutation.cs
// PalindromePermutation.cs

namespace PalindromePermutation.Algo;

public static class PalinPermutation
{
    public static bool CanPermutePalindrome(string s)
    {
        Dictionary<char, int> charCount = [];

        foreach (char c in s)
        {
            if (charCount.ContainsKey(c))
            {
                charCount[c]++;
            }
            else
            {
                charCount[c] = 1;
            }
        }

        int oddCount = 0;
        foreach (int count in charCount.Values)
        {
            if (count % 2 != 0)
            {
                oddCount++;
            }

            if (oddCount > 1)
            {
                return false;
            }
        }

        return true;
    }
}

File: src/e1palindromepermutation/JS/palindromePermutation.js
// palindromePermutation.js

function canPermutePalindrome(s) {
    const charCount = new Map();

    for (const char of s) {
        charCount.set(char, (charCount.get(char) || 0) + 1);
    }

    let oddCount = 0;
    for (const count of charCount.values()) {
        if (count % 2 !== 0) {
            oddCount++;
        }

        if (oddCount > 1) {
            return false;
        }
    }

    return true;
}

module.exports = canPermutePalindrome;
